clients/src/main/java/org/apache/kafka/common/record/LazyDownConvertingRecords.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.record;
+
+import org.apache.kafka.common.utils.Time;
+
+import java.io.IOException;
+import java.nio.channels.GatheringByteChannel;
+
+/**
+ * Records implementation which enables lazy down-conversion to avoid unnecessary memory utilization before
+ * a fetch response is ready to be sent.
+ */
+public class LazyDownConvertingRecords implements Records {
+    private final Records records;
+    private final long firstOffset;
+    private final byte toMagic;
+    private final Time time;
+    private ConvertedRecords<? extends Records> convertedRecords = null;
+
+    public LazyDownConvertingRecords(Records records,
+                                     byte toMagic,
+                                     long firstOffset,
+                                     Time time) {
+        this.records = records;
+        this.toMagic = toMagic;
+        this.firstOffset = firstOffset;
+        this.time = time;
+    }
+
+    @Override
+    public int sizeInBytes() {
+        return convertIfNeeded().records().sizeInBytes();
+    }
+
+    @Override
+    public RecordsSend toSend(String destination) {
+        ConvertedRecords<? extends Records> convertedRecords = convertIfNeeded();
+        return new RecordsSend(destination, convertedRecords.records(), convertedRecords.recordsProcessingStats());
+    }
+
+    @Override
+    public long writeTo(GatheringByteChannel channel, long position, int length) throws IOException {
+        return convertIfNeeded().records().writeTo(channel, position, length);
+    }
+
+    @Override
+    public Iterable<? extends RecordBatch> batches() {
+        return convertIfNeeded().records().batches();
+    }
+
+    @Override
+    public boolean hasMatchingMagic(byte magic) {
+        return convertIfNeeded().records().hasMatchingMagic(magic);
+    }
+
+    @Override
+    public boolean hasCompatibleMagic(byte magic) {
+        return toMagic <= magic;
+    }
+
+    @Override
+    public ConvertedRecords<? extends Records> downConvert(byte toMagic, long firstOffset, Time time) {
+        return records.downConvert(toMagic, firstOffset, time);
```

Comment:
This could potentially block waiting for I/O since it reads all records into memory for down-conversion.

At the moment, this is done on the I/O thread, so the impact of down-conversion is that heap usage is high and the request queue grows longer. Blocking for I/O could delay processing of requests from other connections, but other requests threads can make progress with requests from other connections if one thread is blocked.

If down-conversion is moved to the network thread, the amount of time that we hold onto the extra memory reduces. But we block the network thread for longer while it performs I/O for down-conversion. This prevents any requests from the connections on that network thread to be processed. While it is true that we do I/O currently on the network thread, we don't currently block waiting for I/O on the network thread. 

Would it be possible to modify down-conversion code so that it doesn't block the network thread while waiting for I/O?

Reply:
One option would be to keep a separate queue for fetch responses awaiting down-conversion. We could let the request handlers periodically poll this queue and keep track of the number (or maybe the size) of fetches which have been down-converted and not sent. Say if the number of unsent down-converted fetches is as large as the number of network threads, then we skip polling. That way the conversion is still done in the io threads. What do you think?

Reply:
@hachikuji  Yes, that sounds good to me. It should reduce heap usage when there are lots of requests requiring down-conversions while still avoiding blocking the network thread.